transformations/ocr_perturbation/transformation.py

```diff
@@ -0,0 +1,173 @@
+from tasks.TaskTypes import TaskType
+
+from test.helper import generate_test_cases
+
+
+"""
+This transformation renders input text as an image (optionally augmented with 
+geometric distortions and/or pixel-level noise) and recognizes the rendered text
+using the Tesseract 4 OCR engine. It returns the recognized text as the result.
+"""
+
+tess_lang = {
+    "en": "eng", # English
+    "fr": "fra", # French
+    "es": "spa", # Spanish
+    "de": "deu", # German
+}
+
+trdg_lang = {
+    "en": "en", # English
+    "fr": "fr", # French
+    "es": "es", # Spanish
+    "de": "de", # German
+}
+
+class RenderingParams(object):
+    def __init__(self, background_type=2, distortion_type=0,
+        distortion_orientation=0, blur=0):
+        """
+        :param background_type: 0=noisy, 1=clean, 2=texture, 3=color_texture
+        :param distortion_type: 0=none, 1=sin, 2=cos, 3=random
+        :param distortion_orientation: 0=vertical, 1=horizontal, 2=both
+        :param blur: blur radius
+        """
+        self.background_type = background_type
+        self.distorsion_type = distortion_type
+        self.distortion_orientation = distortion_orientation
+        self.blur = blur
+
+
+class OcrPerturbation(SentenceOperation):
+    tasks = [TaskType.TEXT_CLASSIFICATION, TaskType.TEXT_TO_TEXT_GENERATION]
+    languages = ["en", "fr", "es", "de"]
+    keywords = [
+        "lexical",
+        "morphological",
+        "discourse",
+        "api-based",
+        "tokenizer-required",
+        "unnatural-sounding",
+        "high-precision",
+        "high-coverage",
+        "high-generations",
+    ]
+
+    def __init__(
+        self, seed=0, max_outputs=1, language="en", params=RenderingParams()):
+        """
+        Instatiates an OcrPerturbation object
+        
+        :param seed: random seed
+        :param max_outputs: maximum number of generated results
+        :param language: one of ['en', 'fr', 'es', 'de']
+        :param custom_params: image rendering parameters
+        """
+        assert language in self.languages
+        
+        super().__init__(seed, max_outputs=max_outputs)
+        self.language = language
+        self.params = params
+        self.nlp = self._get_spacy_model()
+
+
+    def generate(self, sentence: str) -> List[str]:
+
+        random.seed(self.seed)
+
+        perturbed_sentences = []        
+        
+        with PyTessBaseAPI(lang=tess_lang[self.language], psm=PSM.RAW_LINE, 
+            oem=OEM.LSTM_ONLY) as ocr:
+            
+            doc = self.nlp(sentence)
+            assert doc.has_annotation("SENT_START")
+
+            image_generators = [self._get_image_generator(sent.text) for sent in doc.sents]
+
+            for idx in range(self.max_outputs):
+            
+                perturbed_sentence = ""
+                    
+                for sent, gen in zip(doc.sents, image_generators):
+
+                    img, lbl = gen.next()
+
```

Comment:
Since lbl is not used, we don't need to store this variable. Replace with underscore.

Reply:
Done.